### PR TITLE
Laplace centrality for issue 4973

### PR DESCRIPTION
**Pull request by gaborberei**
### Reference Issues/PRs

- Related to issue https://github.com/networkx/networkx/issues/4973

### What does this implement/fix? Explain your changes.

- Implementing Laplacian centrality

### Comments?

- Need help with testing

### CODE REVIEW

**Comment by dschult** [comment]
Clicking on the "show all checks" and then one of the failing tests (and then scrolling down through all the testing output to find the error (usually near the bottom)) it shows that the formatting of the centrality values contains either too many digits or not enough. The Examples error checking needs the output to be character-by-character identical to what is produced. 

Try using a similar way of printing the results as that used in [eigenvector.py](https://github.com/networkx/networkx/blob/main/networkx/algorithms/centrality/eigenvector.py). The formatted string determines how many decimal places of accuracy to compare.

**Comment by gaborberei** [comment]
@dschult what do you think about this PR? Is it ready for review?

File: networkx/algorithms/centrality/laplacian.py
Comment (dschult) [inline networkx/algorithms/centrality/laplacian.py:161]:
```suggestion
    laplace_centralities_dict = {}
    for i, node in enumerate(G.nbunch_iter(nbunch)):
```

File: networkx/algorithms/centrality/laplacian.py
Comment (dschult) [inline networkx/algorithms/centrality/laplacian.py:167]:
```suggestion
        # remove row and col i from lap_matrix
        all_but_i = list(np.arange(lap_matrix.shape[0]))
        all_but_i.remove(i)
        A_2 = lap_matrix[all_but_i, :]
        A_2 = A_2[:, all_but_i]
```

File: networkx/algorithms/centrality/laplacian.py
Comment (dschult) [inline networkx/algorithms/centrality/laplacian.py:180]:
```suggestion
        laplace_centralities_dict.update[node] = l_cent
```

File: networkx/algorithms/centrality/laplacian.py
Comment (dschult) [inline networkx/algorithms/centrality/laplacian.py:103]: Update name and spellcheck type.  What is this argument for?

File: networkx/algorithms/centrality/laplacian.py
Comment (dschult) [inline networkx/algorithms/centrality/laplacian.py:98]: ?? I think this is a description of the `nbunch` parameter. Should this be something about normalizing the centrality measure?

File: networkx/algorithms/centrality/laplacian.py
Comment (dschult) [inline networkx/algorithms/centrality/laplacian.py:88]: You should describe what Laplacian Energy is. And in the last line -- avoid "its", perhaps using ``` `G` ```'s

**Review by dschult — COMMENTED** [review]
Thanks for This! 
We need some tests. And the import statements need to be protected from users who don't have numpy and scipy installed. Use the style in `networkx/linalg/laplacian_matrix.py` by putting those imports inside the function. And you'll need to add this function to the list that `needs_numpy` and `needs_scipy` in `networkx/conftest.py`.  Finally, add a link to the documentation in the file `doc/reference/algorithms/centrality.rst`.

I think we shouldn't be creating functions to manipulate the sparse matrices directly. That's for SciPy to do. So I suggested some code that uses indexes and seems to be faster anyway.  There are some other suggestions for simplifying code and/or using networkx functions that already exist.  See what you think.  Feel free to push back on any suggestions in this review.

File: networkx/algorithms/centrality/laplacian.py
Comment (gaborberei) [reply networkx/algorithms/centrality/laplacian.py:103]: Parameters of the nx.directed_laplacian_matrix function. If the users want to specify this for any reason.

File: networkx/algorithms/centrality/laplacian.py
Comment (gaborberei) [reply networkx/algorithms/centrality/laplacian.py:88]: "the sum of squared eigenvalues" is a good description for the Laplacian energy or a more intuitive description is needed?

File: networkx/algorithms/centrality/laplacian.py
Comment (dschult) [reply networkx/algorithms/centrality/laplacian.py]: It looks like the name in the docs doesn't match the name in the function.

File: networkx/algorithms/centrality/laplacian.py
Comment (dschult) [reply networkx/algorithms/centrality/laplacian.py]: That is good for a description of Laplacian energy. :}

File: networkx/algorithms/centrality/laplacian.py
Comment (gaborberei) [reply networkx/algorithms/centrality/laplacian.py]: I resolved the name issue in one of my latest commits. I don't know why these parameters are not available for the undirected graph, I just used the nx.directed_laplacian_matrix and nx.laplacian_matrix functions and for the directed_laplacian there are some extra parameters. I don't know which parameters of the directed laplacian would be important, I just didn't want to restrict the users to use the default specification. I thought if someone wants to specify the directed laplacian matrix parameters they should have the option. But if you think It is not a useful parameter we can get rid of it.

Comment (dschult) [reply networkx/algorithms/centrality/laplacian.py]: Thanks!  I'll look at it ...   I had just forgotten that there was a difference between those function signatures.
Thanks. :}

File: networkx/algorithms/centrality/laplacian.py
Comment (gaborberei) [reply networkx/algorithms/centrality/laplacian.py]: I wrote this here: 

        If True the Laplacian energy of a given node is divided by the Laplacian energy of the entire graph.
        If False the algorithm returns the Laplacian energy of a given node.

**Review by dschult — COMMENTED** [review]
I have some more nit-picky docstring comments. And I have a comment about how the normalization is handled. 

But the big thing missing from this PR are tests.  You need to create a file `tests/test_laplacian.py` (look at the others in that directory for models of how to set up tests with pytest).

**Comment by dschult** [comment]
No... the tests should be in a directory called `tests` inside the directory where the module lives.
So in this case:  `networkx/algorithms/centrality/tests` and the file would be: `networkx/algorithms/centrality/tests/test_laplacian.py`.  sorry for not being more clear the first time.

**Comment by gaborberei** [comment]
@dschult I checked the tests, there is one thing I don't understand. If we have a look at any of the centrality measures test files the concept is something like (this is the example of the [test_closenes_centrality.py](https://github.com/networkx/networkx/blob/main/networkx/algorithms/centrality/tests/test_closeness_centrality.py) ): 

1. define a graph: `K = nx.krackhardt_kite_graph()`
2. calculate the given centrality measure: `c = nx.closeness_centrality(K)`
3. **define the expected output**. `d = {0: 0.529, 1: 0.529, .... , 9: 0.310}`
4. check with an assert statement, if the calculated output is the same as the **defined expected** output. 
  `for n in sorted(K):
                assert c[n] == pytest.approx(d[n], abs=1e-3)`

Here when I **define the expected output** where does it come from? I should calculate it locally or look for examples in different scientific papers? ***How should I define the expected output?***

**Comment by dschult** [comment]
The best approach is to have a simple enough example that you can compute it by hand. But that is not always possible. 
The next best thing is to generate the results locally and add them as a test.  This doesn't actually that that the result is correct, but it checks that it hasn't changed due to some other aspect of the code. We usually try to have at least one test that is simple enough we can generate the true answers by hand. But as stated above, that is not universal in our tests.

**Comment by gaborberei** [comment]
Hi @dschult, It seems like all the tests passed. I have 2 questions regarding the results. First I imported these, and I am not sure what do these mean in case of the testing: 
`np = pytest.importorskip("numpy")`
`sp = pytest.importorskip("scipy")`

Second I couldn't import laplacian_centrality as `nx.laplacian()` I had to import it directly like 
` from networkx.algorithms.centrality.laplacian import laplacian_centrality` which is not the best solution, but at least I could run my tests. 

Could you help me regarding these topics?

**Comment by dschult** [comment]
Q1:  The `pytest.importorskip` function is poorly named. It should be `pytest.import_or_skip`. The idea is that running this command tries to import e.g. `numpy` and if successful, it stores that imported module in the local name `np` or whatever you assign the output of the function to.  If unsuccessful, all the remaining tests in this test file will be skipped by pytest.

So, we use it for modules that require an optional library. If that library is not installed, then these tests are skipped.
Typically, if you don't actually use `np` or `sp` in the testing code, you would not assign the output to a name: `pytest.importorskip("numpy")`.  That ensures that the tests are skipped when needed, and also indicates that the tests themselves don't actually use the numpy library -- though the actual functions they test do.

Q2:  To include the function name in the full name_space, you will need to add it to the `__init__.py` file inside the `centrality` directory. That then gets loaded when `algorithms` runs its `__init__.py` file which gets loaded when networkx runs its `__init__.py` file.

**Comment by gaborberei** [comment]
@dschult, I added `laplacian_centrality` to the `__init__.py`  file, the import works properly. 

What are the next steps? :)

**Comment by dschult** [comment]
I have not had a chance to look through the code in any kind of careful way yet. (2 more weeks of the semester so there is light at the end of the tunnel! :}

I do wonder about the normalized/non-normalized return value. It looks like the normalized returns the % of energy lost when you remove a node. And it looks like the non-normalized option returns the remaining energy after removing the node.  But that is not what the docs say.  It might be better to return "full energy minus after_removal_energy" rather than just the after_removal_energy.  But I might be missing something.

Another fairly big picture question is whether there will be other functions for this module. The name "laplacian.py" is pretty specific and unlikely to attract other measures of centrality. Would this function fit into another module (e.g. eigenvector.py -- though that isn't quite what this is)? Or might there be a larger category of centrality measures that this could fit into? For example: spectral centrality measures?  Are there centrality measures which use the spectrum of a different network matrix?  This really isn't a big deal, but thought I would try to pick your brain a little. :}

I'll try to make a set of actual algorithm/code suggestions soon. Then once I approve it, we'll get another approval and then merge. Thanks for the many tests. It looks like a mix of by-hand and computed examples in there.  What is the test graph "E" and where did it come from?

**Comment by dschult** [comment]
I think I see the graph E in figure 2 of that paper.  That's very helpful to have those published values -- especially if your implementation agrees with them. :}

And now that I looked at the tests more closely, I have a suggestion to be done before merging...
Can you move the tests outside the TestClass structure?  It's something we've been trying to do more of in the last 1-2 years. The test class isn't used in a significant way. The `setup` method is being run before every test and so it has to set up all the graphs for all the tests when it is only running one test.  Can you transition all the tests into test functions instead of test methods? You'll need to build the graph inside the function. But I think that should be easy because you set up the test methods well -- each one uses one graph (I think). That's the hard part. Then it's only a matter of unindenting to make them functions and removing the `class ...` statement near the top. Does this make sense?

**Comment by gaborberei** [comment]
Hi, 

- normalized/non-normalized question: I understand your point. In the original paper I found the following: 
`The importance (centrality) of a vertex v is reflected by the drop of the Laplacian energy of the network to respond to the deactivation (deletion) of the vertex from the network.`
So I think we should follow the rule: the higher the value the more important a node is. Which is `original_energy - after_removal_energy`
- Regarding the `Would this function fit into another module?` question. For now, I would prefer to keep it separately and we can talk about it later. 
- Ok, I can rewrite the test file to use functions instead of one class, this absolutely makes sense.

**Comment by gaborberei** [comment]
Hi @dschult, 

What do you think about the comments and changes? :)

**Comment by gaborberei** [comment]
Hi,  @dschult 

I am back from the summer holiday. I will try to fix the issues this week. What does the "self-assigned" means? Do you have any further suggestions?

**Comment by dschult** [comment]
I tried using the self-assigned feature to make sure I didn't forget about this issue in case you didn't come back after a break. It doesn't have any impact over what you should do -- just a tool for me to keep track of issues that are important to me. Sorry for the confusion.  (Actually I'm just trying it out and it hasn't really worked for me so far... so in that sense it really shouldn't have any impact on what you do with this. :}

File: networkx/algorithms/centrality/laplacian.py
Comment (gaborberei) [reply networkx/algorithms/centrality/laplacian.py]: The problem with this solution is that, if I use the sparse eigenvalue solver I receive this error: 
`TypeError: Cannot use scipy.linalg.eig for sparse A with k >= N - 1. Use scipy.linalg.eig(A.toarray()) or reduce k.`

File: networkx/algorithms/centrality/laplacian.py
Comment (gaborberei) [reply networkx/algorithms/centrality/laplacian.py:10]: I can add these new parameters but isn't it misleading if it only effects the directed laplacian matrix case? That is why I wanted to us the **directed_laplacian_matrix_args solution.

File: networkx/algorithms/centrality/laplacian.py
Comment (dschult) [reply networkx/algorithms/centrality/laplacian.py:115]: I think you can construct a function that both calls `to_array` and also uses the eig function.
For example:
```python
if G.is_directed():
    laplacian_matrix = nx.directed_laplacian_matrix
    eigh = sp.linalg.eigh
else:
    laplacian_matrix = nx.laplacian_matrix
    
    def eigh(A):
        return sp.linalg.eig(A.to_array(), eigvals_only=True)
```
Thoughts?

File: networkx/algorithms/centrality/laplacian.py
Comment (gaborberei) [reply networkx/algorithms/centrality/laplacian.py]: I am note sure I understand this comment. Why I need the full_nodelist object? Could you help me understand this part of the code? 
`nodelist[:] + [n in G if n not in nodeset]`

**Comment by gaborberei** [comment]
Hi @dschult, 

I rewrote the code but I don't understand the current fails. Could you look at this? I Couldn't find the source of the errors. 

And could you help me understand this part of your previous comments? I am just not sure how this part should effect the other parts of the code. 

`nodeset = set(G.nbunch_iter(nodelist))
if len(nodeset) != len(nodelist):
    raise NetworkXError("nodelist contains duplicate nodes or nodes not in G")
full_nodelist = nodelist[:] + [n in G if n not in nodeset]`

**Comment by dschult** [comment]
It looks like a new version of matplotlib in the CI tests is failing,. That is not your PR, it is the new testing configuration. We'll work on that -- you don't need to.  

The first test failuer is an issue with "black" (out code style fixer) It says that `test_laplacian_centrality.py` contains some style changes. You should be able to run `black tests/test_laplacian_centrality.py` locally to fix that -- or if you prefer: `black --diff tests/test_laplacian_centrality.py` to show the differences without making the change automatically.

My suggested code puts the nodes in nodelist as the first rows/columns of the matrix.
```python
nodeset = set(G.nbunch_iter(nodelist)).   ## nbunch_iter drops any nodes not in G. "set" removes duplicates
if len(nodeset) != len(nodelist):        ## if we lost any nodes then nodelist isn't good...
    raise NetworkXError("nodelist contains duplicate nodes or nodes not in G")
full_nodelist = nodelist[:] + [n in G if n not in nodeset]. ## now add the nodes from G at the end of the list.
```
That puts the nodelist nodes first with the other nodes at the end. The ordering in the matrices is determined that way.

**Comment by rossbar** [comment]
It looks like you've accidentally committed an `env/` directory which has added 500+ files. I suspect this is the source of the errors in CI.

**Comment by gaborberei** [comment]
Thanks for your help! :) 

It wasn't **black**, but **isort** which generated the **style** fail. I also deleted the env files. Now working on the last comment.

**Comment by gaborberei** [comment]
Hi @dschult, 

I implemented the comments, the tests are passed. Please have a look at the code if you have some time. Thanks for your help.

**Comment by gaborberei** [comment]
Hi @dschult, 

what do you think about the code? Did you have some time to check it?

**Review by dschult — COMMENTED** [review]
I think this is ready to include.  
Can you merge the current nx main branch so we can get the documentation CI test working and check that out before merging?
Thanks!

File: networkx/algorithms/centrality/laplacian.py
Comment (rossbar) [inline networkx/algorithms/centrality/laplacian.py:89]: Just a nit, but IMO the code is easier to read without this indirection. Since this is just a one-liner, I think things become more readable if you replace all the instances of `eigh_f` with `sp.linalg.eigh(A.toarray(), eigvals_only=True)`.

File: networkx/algorithms/centrality/laplacian.py
Comment (rossbar) [inline networkx/algorithms/centrality/laplacian.py:108]: NetworkX now prefers SciPy's sparse arrays to sparse matrices, so this wrapper should be removed (sparse arrays are returned by default for NX 3.0).

Note this may also necessitate updating indexing semantics, e.g. in `new_diag` below.

File: networkx/algorithms/centrality/laplacian.py
Comment (rossbar) [inline networkx/algorithms/centrality/laplacian.py:105]: I think using a set here may cause problems - the `nodelist` parameter description in the docstring explicitly mentions that it determines node ordering, but sets do not preserve order.

If node ordering actually isn't important, then I'd vote to update the docstring. Either way, I think the else branch could be simplified to `full_nodelist = list(G)`.

File: networkx/algorithms/centrality/laplacian.py
Comment (rossbar) [inline networkx/algorithms/centrality/laplacian.py:115]: Just a nit, but since you compute the eigenvalues in *both* branches of the conditional, might as well move it outside:

```suggestion
    else:
        lap_matrix = nx.laplacian_matrix(G, full_nodelist, weight)
    
    eigh = sp.linalg.eigh(lap_matrix.toarray(), eigvals_only=True)
```

File: networkx/algorithms/centrality/laplacian.py
Comment (rossbar) [inline networkx/algorithms/centrality/laplacian.py:132]: Another nit, but might as well combine the row/column advanced indexing operations into one line:

```suggestion
        A_2 = lap_matrix[all_but_i, :][:, all_but_i]
```

File: networkx/algorithms/centrality/laplacian.py
Comment (rossbar) [inline networkx/algorithms/centrality/laplacian.py:134]: Similarly this can be simplified with advanced indexing.

```suggestion
        A_2.setdiag(new_diag[all_but_i])
```

**Review by rossbar — COMMENTED** [review]
Just a couple comments from a quick scan of the function. I'll try to make a more detailed pass in the near future!

File: networkx/algorithms/centrality/laplacian.py
Comment (dschult) [reply networkx/algorithms/centrality/laplacian.py:105]: In the code block for the `if` clause (nodelist provided) the nodeset should be fine because it is only used to check for duplicates and nodes in G. It does not impact the order of the resulting `full_nodeset`.

The `else` clause should definitely be simplified to `full_nodelist(G)`.   Thanks!

Comment (gaborberei) [reply networkx/algorithms/centrality/laplacian.py:105]: The purpose of the `nodelist` variable is not only checking for duplicates and nodes in G, but the `laplacian_centrality` is calculated for the nodes in the `nodelist`. You can see it at the beginning of the for loop. If `nodelist` is None that means someone wants to calculate the centrality for all the nodes. 

As the definition of this metric is:

"""The Laplacian Centrality of a node `i` is measured by the drop in the Laplacian Energy after deleting node `i` from the graph."""

First I have to calculate the energy of the entire graph and then I can calculate the metric for the nodes in `nodelist` or for the entire graph (if `nodelist =None`)

File: networkx/algorithms/centrality/laplacian.py
Comment (gaborberei) [reply networkx/algorithms/centrality/laplacian.py:108]: As I see only the NX  2.8.7 is available and NX 3 dev version. Please correct me if I am wrong, but as I see it is possible to merge the current version of this code into the current version of NX and then I can rewrite the code and merge the new version to the NX 3? 

I think this can be a viable option as this feature will be available in both versions of NX. What do you think @rossbar ?

**Comment by gaborberei** [comment]
Hi @dschult, @rossbar 

what do you think about the comments?

File: networkx/algorithms/centrality/laplacian.py
Comment (dschult) [reply networkx/algorithms/centrality/laplacian.py:102]: The case when nodelist is None the code does not need to check whether nodes are in the graph, or duplicate. It should just set nodelist and full_nodelist to list of all the nodes in the graph. It does not need to be a set. And there is a minor memory savings if it uses the same object for the nodelist and the full_nodelist. So, I think the code should be something like:
```python
nodelist = full_nodelist = list(G)
```
And then replace the only place where nodeset is used with `nodelist`.  Note that this will work when nodelist is not None as well.  The only reason we should be creating a set here is to identify duplicate nodes in a provided nodelist.

File: networkx/algorithms/centrality/laplacian.py
Comment (dschult) [reply networkx/algorithms/centrality/laplacian.py:105]: sparse arrays are preferred now -- We don't need to wait for NX3.0.

**Comment by dschult** [comment]
I went ahead and made some changes that had been discussed. And that led to some that hadn't been discussed.

Discussed stuff:
- changed nodelist preparation code
- remove casting as csr_matrix. All computation now done with ndarrays since we need dense for eigh anyway. This makes `getcol` available and used `fill_diagonal` for setting the diag entries.

Other stuff:
- The normalized code was correct, but the non-normalized version was computing `1 - new_energy` rather than `full_energy - new_energy`.  This fix simplified the code somewhat too.
- No tests were present for normalized=False and weight=None. Added some.
- wrapped the doc_strings to same as what `black` would provide. Reworded some.
- other minor changes for error msgs and spacing.

Feel free to push back on any of this. Make a push to the branch, or comment here in the PR.

**Comment by gaborberei** [comment]
I think it is fine, thanks. 

What else we need?

**Review by dschult — APPROVED** [review]
Looks good to me!
:}

**Comment by gaborberei** [comment]
Great, then, what's next? :)

**Comment by dschult** [comment]
We get a second approval (set of eyes) and then we can merge. :}

**Comment by gaborberei** [comment]
Hi @rossbar, please could you look at the code? 

It would be the best present for Christmas if the code was approved! :)

**Comment by gaborberei** [comment]
Hi @dschult, 

Do I have to do anything now? Do I have to accept the changes as well? Or just chill and wait for the second approval?

**Comment by gaborberei** [comment]
Hi @dschult , @rossbar 

I hope you have a great new year so far. Could you check the changes? Do I have to do anything with the code? What do you think, is it ready for merge?

**Pull request merged**
